File: packages/metame-contracts/src/types/surfacePlan.ts
import { z } from "zod";
import { SurfaceSchema, DensitySchema, DeviceClassSchema, OrientationSchema, InteractionTypeSchema, RealEstateSchema } from "./contentModuleRenderProfile";

// Reference types
export const RefKindSchema = z.enum(["schema_ref", "doc_ref", "uri_ref"]);
export type RefKind = z.infer<typeof RefKindSchema>;

export const RefSchema = z.object({
  kind: RefKindSchema,
  id: z.string(),
});
export type Ref = z.infer<typeof RefSchema>;

// Device context
export const DeviceContextSchema = z.object({
  device_class: DeviceClassSchema,
  orientation: OrientationSchema,
  interaction: InteractionTypeSchema,
  real_estate: RealEstateSchema,
});
export type DeviceContext = z.infer<typeof DeviceContextSchema>;

// Intent
export const IntentSchema = z.object({
  user_ask: z.string(),
  mode: z.enum(["be", "make", "play", "earn", "share"]),
  focus: z.optional(z.string()),
});
export type Intent = z.infer<typeof IntentSchema>;

// Module reference
export const ModuleRefSchema = z.object({
  module_id: z.string(),
  module_type: z.string(),
  render_profile_ref: RefSchema,
  source_refs: z.optional(RefSchema.array()),
});
export type ModuleRef = z.infer<typeof ModuleRefSchema>;

// Placement overrides
export const PlacementOverridesSchema = z.object({
  max_lines: z.optional(z.number()),
  collapse_sections: z.optional(z.boolean()),
  hide_media: z.optional(z.boolean()),
});
export type PlacementOverrides = z.infer<typeof PlacementOverridesSchema>;

// Placement interaction
export const PlacementInteractionSchema = z.object({
  opens: SurfaceSchema,
  open_density: DensitySchema,
});
export type PlacementInteraction = z.infer<typeof PlacementInteractionSchema>;

// Placement
export const PlacementSchema = z.object({
  module_id: z.string(),
  surface: SurfaceSchema,
  density: DensitySchema,
  region: z.enum(["primary", "secondary", "footer", "header", "sidebar", "canvas"]),
  order: z.number(),
  interaction: z.optional(PlacementInteractionSchema),
  overrides: z.optional(PlacementOverridesSchema),
  reasoning_tags: z.optional(z.string().array()),
});
export type Placement = z.infer<typeof PlacementSchema>;

// Navigation
export const NavigationSchema = z.object({
  entry_surface: SurfaceSchema,
  progression: SurfaceSchema.array(),
});
export type Navigation = z.infer<typeof NavigationSchema>;

// Verification references
export const VerificationRefsSchema = z.object({
  dis_ref: RefSchema,
  constraint_manifest_ref: RefSchema,
  parity_report_ref: RefSchema,
});
export type VerificationRefs = z.infer<typeof VerificationRefsSchema>;

// Audit references
export const AuditRefsSchema = z.object({
  trace_id: z.string(),
  span_id: z.string(),
  actor: z.string(),
  event_hashes: z.string().array(),
});
export type AuditRefs = z.infer<typeof AuditRefsSchema>;

// Main surface plan schema
export const SurfacePlanV0Schema = z.object({
  schema_version: z.literal("0.1.0"),
  plan_id: z.string(),
  session_id: z.string(),
  cartridge: z.string(),
  codex_id: z.optional(z.string()),
  capsule_id: z.optional(z.string()),
  thread_id: z.optional(z.string()),
  intent: IntentSchema,
  device_context: DeviceContextSchema,
  modules: ModuleRefSchema.array(),
  placements: PlacementSchema.array(),
  navigation: NavigationSchema,
  verification: VerificationRefsSchema,
  audit: z.optional(AuditRefsSchema),
});
export type SurfacePlanV0 = z.infer<typeof SurfacePlanV0Schema>;

File: services/metame_runtime/models/surface_plan.py
from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional
from enum import Enum
from pydantic import BaseModel, Field

from .content_module_render_profile import Surface, Density, DeviceClass, Orientation, InteractionType, RealEstate


class RefKind(str, Enum):
    schema_ref = "schema_ref"
    doc_ref = "doc_ref"
    uri_ref = "uri_ref"


class Ref(BaseModel):
    kind: RefKind
    id: str


class DeviceContext(BaseModel):
    device_class: DeviceClass
    orientation: Orientation
    interaction: InteractionType
    real_estate: RealEstate


class Intent(BaseModel):
    user_ask: str
    mode: Literal["be", "make", "play", "earn", "share"]
    focus: Optional[str] = None


class ModuleRef(BaseModel):
    module_id: str
    module_type: str
    render_profile_ref: Ref
    source_refs: Optional[List[Ref]] = None


class PlacementOverrides(BaseModel):
    max_lines: Optional[int] = None
    collapse_sections: Optional[bool] = None
    hide_media: Optional[bool] = None


class PlacementInteraction(BaseModel):
    opens: Surface
    open_density: Density


class Placement(BaseModel):
    module_id: str
    surface: Surface
    density: Density
    region: Literal["primary", "secondary", "footer", "header", "sidebar", "canvas"]
    order: int
    interaction: Optional[PlacementInteraction] = None
    overrides: Optional[PlacementOverrides] = None
    reasoning_tags: Optional[List[str]] = None


class Navigation(BaseModel):
    entry_surface: Surface
    progression: List[Surface]


class VerificationRefs(BaseModel):
    dis_ref: Ref
    constraint_manifest_ref: Ref
    parity_report_ref: Ref


class AuditRefs(BaseModel):
    trace_id: str
    span_id: str
    actor: str
    event_hashes: List[str]


class SurfacePlanV0(BaseModel):
    schema_version: Literal["0.1.0"]
    plan_id: str
    session_id: str
    cartridge: str
    codex_id: Optional[str] = None
    capsule_id: Optional[str] = None
    thread_id: Optional[str] = None
    intent: Intent
    device_context: DeviceContext
    modules: List[ModuleRef]
    placements: List[Placement]
    navigation: Navigation
    verification: VerificationRefs
    audit: Optional[AuditRefs] = None
